packages/devtools_shared/lib/src/utils/file_utils.dart

```diff
@@ -2,20 +2,88 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:dtd/dtd.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+
 /// Attempts to detect the package root of [fileUriString].
 ///
-/// This method uses the heuristics that a Dart executable within a Dart package
-/// should reside in one of the following top-level folders: 'lib', 'bin',
-/// 'integration_test', 'test', or 'benchmark'.
-String packageRootFromFileUriString(String fileUriString) {
-  // TODO(kenz): for robustness, consider sending the root library uri to the
-  // server and having the server look for the package folder that contains the
-  // `.dart_tool` directory.
+/// This method first tries to use the Dart Tooling Daemon to walk the
+/// directory structure of [fileUriString] and look for the package root; we
+/// consider a directory to be a package root if it contains the `.dart_tool`
+/// directory.
+///
+/// If we cannot find the package root using the Dart Tooling Daemon, we use the
+/// heuristic that a Dart executable within a Dart package should reside in one
+/// of the following top-level folders:
+///
+/// * lib
+/// * bin
+/// * integration_test
+/// * test
+/// * benchmark
+/// * example
+///
+/// The URI returned will be a file URI String and will NOT have a trailing
+/// slash.
+Future<String> packageRootFromFileUriString(
+  String fileUriString, {
```

Comment:
Is this always assumed to be a `file://` URI? Do we know it is? `Uri.parse` will interpret Windows drive letters as the scheme if it isn't prefixed with `file://`.

Reply:
It is expected to be a file URI. I can add an assert at the top of this method to make sure.

Reply:
I added an assertion and a test case.